cmd/object-handlers.go

```diff
@@ -1757,8 +1757,11 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 		md5hex = "" // Do not try to verify the content.
 		sha256hex = ""
 	}
-
-	hashReader, err := hash.NewReader(reader, size, md5hex, sha256hex, actualSize)
+	// SSE-KVM and SSE-C if not request Content-Md5. Simple use md5.Sum(uuid) as etag.
+	if !etag.ContentMD5Requested(r.Header) && (crypto.S3KMS.IsRequested(r.Header) || crypto.SSEC.IsRequested(r.Header)) {
+		ctx = context.WithValue(ctx, etag.ContextEtagAsUUID, mustGetUUID())
```

Comment:
We shouldn't start passing functional parameters as context values.

If we do that, we will end up with a huge amount of magic code that does things that are hard to follow.

Pass a regular parameter. You are welcome to add a `type hash.ReaderOptions struct` that combines the parameters, but you can also just add another one.

Reply:
Yeah I thought about it but it may or may not be okay here. But I agree we can change the function signature we are changing it anyways change it to take

```
opts := hash.ReaderOptions{
}

hash.NewReader(ctx, r.Body, opts)
```

Is perhaps a more idiomatic approach @jiuker